### DISCUSSION TITLE
How to annotate class instance variable / bound method?

### DISCUSSION THREAD

**Discussion by cuu508** [python/typing#1648] [Q&A] [answered]
Suppose I have a class that allows a field to be either a boolean or a bound method returning a boolean:

```python
from __future__ import annotations

from typing import Callable


class Parent(object):
    foo: Callable[[], bool] | bool

    def print_foo(self):
        print(self.foo() if callable(self.foo) else self.foo)


class Child(Parent):
    def foo(self) -> bool:
        return True


class Child2(Parent):
    foo = False


Child().print_foo()
Child2().print_foo()
# prints:
# True
# False
```

In the above example, the subclass `Child` defines `foo` as a method, but the subclass `Child2` declares it as a a variable. The code works, but mypy is not happy with the type annotation for foo:

```
experiment.py:14: error: Signature of "foo" incompatible with supertype "Parent"  [override]
experiment.py:14: note:      Superclass:
experiment.py:14: note:          Callable[[], bool] | bool
experiment.py:14: note:      Subclass:
experiment.py:14: note:          def foo(self) -> bool
Found 1 error in 1 file (checked 1 source file)
```

If I remove the union, and allow foo to only be a bound method, mypy is happy:

```python
# This works
class Parent(object):
    foo: Callable[[], bool] 


class Child(Parent):
    def foo(self) -> bool:
        return True
```

But as soon as I add `| bool` back to the type annotation, I get a mypy error.

What am I doing wrong, and how can I fix the type annotation?

For context, this is a toy example, the actual issue is in django-stubs here: https://github.com/typeddjango/django-stubs/issues/1755

**Comment by Akuli** [comment, marked as answer]
I think the method isn't compatible with `foo: Callable[[], bool]`, because that means `foo` is a read-write attribute, but mypy thinks of methods as read-only. For example, `Child().foo = lambda: True` gives an error. So `Child` doesn't have all of the functionality (writing `foo`) that every `Parent` instance is supposed to have.

Defining `foo` as a read-only attribute using `@property` works for me:

```python
from typing import Callable, ClassVar

class Parent(object):
    @property
    def foo(self) -> Callable[[], bool] | bool:
        raise NotImplementedError

    def print_foo(self):
        print(self.foo() if callable(self.foo) else self.foo)

class Child(Parent):
    def foo(self) -> bool:
        return True

class Child2(Parent):
    foo = False

Child().print_foo()
Child2().print_foo()
```

This seems to be somewhat mypy-specific. If I check the same code with pyright, I get errors.
